### PR TITLE
Converter: implement conversion framework

### PR DESCRIPTION
**Pull request by imeoer**
Introducing a standalone acceleration image conversion service designed to
provide Harbor with the ability to automatically convert user images to
accelerated images. When a user does something such as artifact push,
Harbor will request the service to complete the corresponding image
conversion through its integrated Nydus, eStargz, etc. drivers.

Currently only the nydus driver is implemented, the estargz driver will
be integrated next.

Signed-off-by: Yan Song <imeoer@linux.alibaba.com>

### CODE REVIEW

File: pkg/content/content.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+	return nil
+}
+
+func (pvd *LocalContentProvider) Write(ctx context.Context, desc ocispec.Descriptor, reader io.Reader) error {
+	digest := desc.Digest.String()
+
+	gcLabels := map[string]string{
+		"containerd.io/gc.ref.content": digest,
+	}
```
Comment (ktock) [inline pkg/content/content.go:118]: Why does this GC label point to the blob itself?

File: pkg/content/content.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+	descs := []ocispec.Descriptor{manifest.Config}
+	descs = append(descs, manifest.Layers...)
+
+	// Push the target image config and all layers to remote registry.
+	if err := images.Dispatch(ctx, pushHandler, nil, descs...); err != nil {
+		return errors.Wrapf(err, "push config and layer")
+	}
+
+	// Push the manifest of target image to remote registry.
+	return images.Dispatch(ctx, pushHandler, nil, desc)
```
Comment (ktock) [inline pkg/content/content.go:176]: Why not use [Push API](https://github.com/containerd/containerd/blob/483e23bdcdf46ae204f267d6864d1a51da7ec1ae/client.go#L421)?

File: pkg/converter/converter.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+func (cvt *LocalConverter) Dispatch(ctx context.Context, ref string, sync bool) error {
+	if sync {
+		// FIXME: The synchronous conversion task should also be
+		// executed in a limited worker queue.
+		return cvt.Convert(context.Background(), ref)
+	}
+
+	cvt.worker.Dispatch(func() error {
+		return metrics.Conversion.OpWrap(func() error {
+			return cvt.Convert(context.Background(), ref)
```
Comment (ktock) [inline pkg/converter/converter.go:141]: The number of worker is finite so we need a timeout maybe (can be following-up patches though)

File: pkg/content/content.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+	Pull(ctx context.Context, ref string) error
+	// Push pushes target image to remote registry by specified reference,
+	// the desc parameter represents the manifest of targe image.
+	Push(ctx context.Context, desc ocispec.Descriptor, ref string) error
+
+	// Write writes a blob to containerd content store, the blob can be an
+	// image manifest, index, layer or config.
+	Write(ctx context.Context, desc ocispec.Descriptor, reader io.Reader) error
+	// Read reads a blob from containerd content store.
+	Read(ctx context.Context, desc ocispec.Descriptor) ([]byte, error)
```
Comment (ktock) [inline pkg/content/content.go:52]: `content.Store` should be exposed directly.

**Review by ktock — COMMENTED** [review]
Thank you for working on this! This patch worked on my machine.

**Comment by imeoer** [comment]
@steven-zou Can we add maintain or write type access role for @ktock ?

**Comment by imeoer** [comment]
@steven-zou Thank you for adding the access role.
cc @ktock You should already have review permission :).

File: pkg/content/content.go (hunk 1 above)
Comment (imeoer) [reply pkg/content/content.go:130]: Ok, I will take a look the usage of GC label.

**Comment by ktock** [comment]
Thanks @imeoer and @steven-zou !

File: pkg/content/content.go (hunk 1 above)
Comment (imeoer) [reply pkg/content/content.go]: Moved GC label to the manifest blob, does the config or layer blob itself need to be added GC label? I haven't fully understood when containerd will reclaim a blob that has just been written.

File: pkg/content/content.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+func (pvd *LocalContentProvider) Pull(ctx context.Context, ref string) error {
+	refUrl, err := url.Parse(fmt.Sprintf("dummy://%s", ref))
+	if err != nil {
+		return errors.Wrap(err, "parse reference of source image")
+	}
+	var resolver remotes.Resolver
+	if auth, ok := pvd.cfg.Source[refUrl.Host]; ok {
+		resolver = remote.NewResolver(auth.Insecure, remote.NewBasicAuthCredFunc(auth.Auth))
+	} else {
+		resolver = remote.NewResolver(false, remote.NewDockerConfigCredFunc())
```
[the remark this answers is not in the document]
Comment (imeoer) [reply pkg/content/content.go]: Fixed

File: README.md [hunk 6]
```diff
@@ -1,2 +1,93 @@
...
+
+2. Edit config file
+
+Following comments in `misc/config/config.yaml`, ensure base64 encoded auth string and webhook auth header be configured correctly.
+
+### Boot service
+
+1. Ensure containerd service is running
+
+Acceleration service depends on containerd service to manage image content, containerd service is using default if you have installed the latest docker.
```
[the remark this answers is not in the document]
Comment (imeoer) [reply README.md]: Fixed

File: .gitignore [hunk 7]
```diff
@@ -14,4 +14,5 @@
 # Dependency directories (remove the comment below to include it)
 # vendor/
 
+/misc/config/config.yaml
```
Comment (ktock) [inline .gitignore:17]: Why is this ignored?

File: Makefile [hunk 8]
```diff
@@ -2,5 +2,20 @@ GIT_COMMIT := $(shell git rev-list -1 HEAD)
 BUILD_TIME := $(shell date -u +%Y%m%d.%H%M)
 CWD := $(shell dirname $(realpath $(firstword $(MAKEFILE_LIST))))
 
+# Build binary to ./acceleration-service
 default:
+	go vet ./pkg/...
```
Comment (ktock) [inline Makefile:7]: Could we use golangci-lint?

File: pkg/content/content.go [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+	Pull(ctx context.Context, ref string) error
+	// Push pushes target image to remote registry by specified reference,
+	// the desc parameter represents the manifest of targe image.
+	Push(ctx context.Context, desc ocispec.Descriptor, ref string) error
+
+	// Write writes a blob to containerd content store, the blob can be an
+	// image manifest, index, layer or config.
+	Write(ctx context.Context, desc ocispec.Descriptor, reader io.Reader, labels map[string]string) error
+	// Read reads a blob from containerd content store.
+	Read(ctx context.Context, desc ocispec.Descriptor) ([]byte, error)
```
Comment (ktock) [inline pkg/content/content.go:51]: Now that we have `ContentProvider.ContentStore()` API, so could we remove `Write` and `Read` APIs and let the caller to use `ContentStore` instead?

File: Makefile (hunk 8 above)
Comment (imeoer) [reply Makefile]: Added, thanks for the recommendation.

File: misc/config/config.yaml.tmpl [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  worker: 5
+  driver:
+    # accelerator driver type: `nydus`
+    # estargz driver implementation is in progress
+    type: nydus
+    config:
+      work_dir: /tmp
+      # `nydus-image` binary path, download it from:
+      # https://github.com/imeoer/image-service/releases/download/v1.0.0/nydus-image
+      builder: nydus-image
```
Comment (ktock) [inline misc/config/config.yaml.tmpl:43]: Is it possible to enable multiple drivers? Do we need to run multiple acceleration-service processes with different configurations to enable multiple converters?

Comment (imeoer) [reply misc/config/config.yaml.tmpl:43]:
``` yaml
  drivers:
    - type: nydus
      config:
        # ...
      rules:
          # source image reference that matches the regex will be converted
        - match: ns1/*
          tag_suffix: -converted
    - type: estargz
      config:
        # ...
      rules:
        - match: ns2/*
          tag_suffix: -converted
```

Should be a possible scenario (it's not difficult to support), how about this format? The source image will be converted by the first matched rule.

**Comment by imeoer** [comment]
> @imeoer How many LGTMs are needed to merge?

@ktock It seems that need only one :).

**Comment by ktock** [comment]
eStargz converter will be added in https://github.com/goharbor/acceleration-service/pull/2.

**Pull request merged**